aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,238 @@
+     *
+     * @throws Exception if upload is not canceled successfully
+     *         before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsCancelable() throws Exception {
+        final CountDownLatch canceled = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> fail("Upload finished before being successfully cancelled."),
+            onError -> fail("Upload failed for a different reason.")
+        );
+
+        // Listen to Hub events to cancel when progress has been made
+        SubscriptionToken progressToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadProgress".equals(hubEvent.getName())) {
+                HubEvent<Float> progressEvent = (HubEvent<Float>) hubEvent;
+                Float progress = progressEvent.getData();
+                if (progress != null && progress > 0) {
+                    op.cancel();
+                }
+            }
+        });
+        subscriptions.add(progressToken);
+
+        // Listen to Hub events for cancel
+        SubscriptionToken cancelToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadState".equals(hubEvent.getName())) {
+                HubEvent<String> stateEvent = (HubEvent<String>) hubEvent;
+                TransferState state = TransferState.getState(stateEvent.getData());
+                if (TransferState.CANCELED.equals(state)) {
+                    canceled.countDown();
+                }
+            }
+        });
+        subscriptions.add(cancelToken);
+
+        // Assert that the required conditions have been met
+        assertTrue(canceled.await(EXTENDED_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS));
+    }
+
+    /**
+     * Tests that file upload operation can be paused and resumed
+     * while the transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not paused, resumed, and
+     *         completed successfully before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsResumable() throws Exception {
+        final CountDownLatch completed = new CountDownLatch(1);
+        final CountDownLatch resumed = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> completed.countDown(),
+            onError -> fail("Upload is not successful.")
+        );
+
+        // Listen to Hub events to pause when progress has been made
+        SubscriptionToken pauseToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadProgress".equals(hubEvent.getName())) {
```

Comment:
For datastore, I created a single file where I keep all of the event names that are possibly published by datastore. I think we should try to do the same for the different categories.

Also - we need to figure out how we are going to organize *category* events vs. *plugin* events. And how to namespace them!! Anyway, I would at least put this `uploadProgress` string into an enum called like `StorageCategoryEventNames` (or whatever is appropriate to match the existing version of this for DataStore.)